Constitutional AI: Harmlessness from AI Feedback

gleibovich-nvidia opened the pull request:
Draft for the implementation of  [Constitutional AI: Harmlessness from AI Feedback](https://arxiv.org/abs/2212.08073). 

Note that there are still some minor missing pieces regarding the exact model and datasets to be used, yet the main building blocks are in place.

terrykong commented on examples/nlp/cai/generate_rl_cai_dataset.py:
```diff
@@ -0,0 +1,741 @@
+        resp_sentences = resp_json["sentences"]
+        return resp_sentences
+
+    data = {
+        "sentences": prompt,
+        "tokens_to_generate": 1024,
+        "temperature": temperature,
+        "add_BOS": True,
+        "top_k": 50,
+        "top_p": 0.95,
+        "greedy": False,
+        "all_probs": False,
+        "repetition_penalty": 1,
+        "min_tokens_to_generate": 1,
+        "end_strings": ["<extra_id_1>"],
+    }
```
Can you please parametrize this json obj? If someone wants to try generating 2048 tokens, it would be helpful to not have to edit code. 

hydra may be too heavy for this script, so perhaps plumbing these via [add_argument_group](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_argument_group) would be a good middle ground.

snisimov replied:
Thank you for your suggestion.
I've implemented the **add_argument_group** approach to parameterize the JSON object as you recommended.

snisimov replied:
Thanks, done.
we have implemented the use of 'add_argument_group' to parameterize the JSON object as you recommended.

terrykong commented on examples/nlp/cai/generate_rl_cai_dataset.py:
```diff
@@ -0,0 +1,741 @@
+    payload = {
+        "messages": messages,
+        "temperature": temperature + 0.0000001,
+        "top_p": 0.0000001,
+        "max_tokens": 1024,
```
Same comment about parametrizing these arguments

terrykong commented on examples/nlp/cai/generate_rl_cai_dataset.py:
```diff
@@ -0,0 +1,741 @@
+    turn_token = "<extra_id_1>"  # (model_config: chat_prompt_tokens.turn_start)
+    end_signal = "\n"  # (model_config: chat_prompt_tokens.end_of_turn. NOTE: "\x0A" is '\n' in ASCII code)
+    label_start = "<extra_id_2>"  # (model_config: chat_prompt_tokens.label_start)
+    end_name_signal = "\n"  # (model_config: chat_prompt_tokens.end_of_name. NOTE: "\x0A" is '\n' in ASCII code)
```
Can we parametrize these and reuse the variable holding them throughout this script. `extra_id_*` is hardcoded in many places and it would be cleaner to reference a variable. WDYT about plumbing it from the parser?

snisimov replied:
Thanks, done.
we have implemented the use of 'add_argument_group' to parameterize the variables.

terrykong commented on examples/nlp/cai/generate_sl_cai_dataset.py:
```diff
@@ -0,0 +1,484 @@
+        return resp_sentences
+
+    data = {
+        "sentences": prompt,
+        "add_BOS": True,
+        "top_k": 1,
+        "top_p": 0.9,
+        "all_probs": False,
+        "repetition_penalty": 1.2,
+        "min_tokens_to_generate": 1,
```
WDYT about parametrizing this?

snisimov replied:
Thank you for your suggestion.
I've implemented the add_argument_group approach to parameterize the JSON object as you recommended.
i will do the same for 'examples/nlp/cai/generate_rl_cai_dataset.py'

terrykong commented on examples/nlp/cai/process_anthropic_hh_using_chat_prompt.py:
```diff
@@ -0,0 +1,243 @@
+        string_to_use = string_to_use.rstrip("\n") + "\n"
+
+        if add_eos:
+            string_to_use += "<extra_id_1>"
```
Can this be parametrized?

snisimov replied:
Thanks, done.

snisimov commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+           --output-file-name-prefix anthropic_helpful_only
+
+
+Step 2: Generate and revise responses to harmfulness prompts creating the SL-CAI dataset
```
done

snisimov commented on examples/nlp/cai/generate_rl_cai_dataset.py:
```diff
@@ -0,0 +1,741 @@
+    return all_samples
+
+
+def generate_responses_batch(prompt_list: list, temperature: int, port_num: int):
```
fixed

snisimov commented on examples/nlp/cai/generate_sl_cai_dataset.py:
```diff
@@ -0,0 +1,484 @@
+
+    if args.max_seq_length is not None:
+        assert args.max_seq_length > 0
+        assert os.path.isfile(args.tokenizer_model)
```
done

snisimov commented on examples/nlp/cai/generate_rl_cai_dataset.py:
```diff
@@ -0,0 +1,741 @@
+    headers = {"Content-Type": "application/json"}
+
+    def request_data(request):
+        resp = requests.put("http://localhost:{}/generate".format(port_num), data=json.dumps(request), headers=headers)
```
The service needs to be started by the user. I've added an argument to the parser to specify the hostname or IP address of the inference service, which allows the user to define where the service is running.

gleibovich-nvidia commented on examples/nlp/cai/generate_rl_cai_dataset.py:
```diff
@@ -0,0 +1,741 @@
+        return response
+
+
+constitution = """
```
Good point that we definitely need to discuss. 
As CAI paper is 1.5 years old by now [= forever in LLM developement :-)] and has no official code implementation, we agreed with @gshennvm that we will roughly follow the paper and might make some changes where we see fit. Given that the supported context size for recent models is getting increasingly larger each day, we experimented with feeding the entire constitution to the judge LLM and have it judge according to all the principles at once. Empirically we found that working quite well, as the selected most harmless and most harmful responses made perfect sense on several tens of prompts that we experimented with. 

What do you think? @terrykong @gshennvm 

P.S: there are other section which is not exactly following the paper, and we might want to change those as well if we want to follow the paper exactly.  e.g. in the paper in order to rank and choose between the responses, they use the logprob of each of the responses as the criteria for selection, whereas in our implementation we use an external judge LLM as, for instance, is being done with MT-Bench for scoring LLMs.

terrykong replied:
I personally think it's fine to deviate from the algorithm as long as it's documented; at the very least in a comment.

I do think the constitutions should be passed into the script as opposed to being a literal within the python script. It makes customization more difficult since we are mixing "data" with code

gleibovich-nvidia replied:
Absolutely agree. Will fix and make sure to properly document! 
Thanks for the feedback.

gleibovich-nvidia replied:
Done. Constitution is now in a separate file and paper deviation is properly documented.

snisimov commented on examples/nlp/cai/generate_sl_cai_dataset.py:
```diff
@@ -0,0 +1,484 @@
+    headers = {"Content-Type": "application/json"}
+
+    def request_data(request):
+        resp = requests.put("http://localhost:{}/generate".format(port_num), data=json.dumps(request), headers=headers)
```
The service needs to be started by the user. I've added an argument to the parser to specify the hostname or IP address of the inference service, which allows the user to define where the service is running.

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+
+.. code-block:: bash
+
+   python /opt/NeMo/examples/nlp/language_modeling/megatron_gpt_eval.py 
```
done

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+
+.. code-block:: bash
+
+   python examples/nlp/cai/generate_sl_cai_dataset.py 
```
done

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+
+.. code-block:: bash
+
+   python examples/nlp/gpt/train_gpt_sft.py
```
done

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+
+.. code-block:: bash
+
+   python /opt/NeMo/examples/nlp/language_modeling/megatron_gpt_eval.py
```
done

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+
+.. code-block:: bash
+
+   python examples/nlp/cai/generate_rl_cai_dataset.py
```
done

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+.. image:: ../assets/cai_diagram.png
+   :alt: basic steps of the CAI process
+
+Both the critiques and the AI feedback are steered by a small set of principles drawn from a ‘constitution’. The supervised stage significantly improves the initial model, and gives some control over the initial behavior at the start of the RL phase, addressing potential exploration problems. The RL stage significantly improves performance and reliability.
```
done

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+
+(Supervised Stage) Critique → Revision → Supervised Learning: The AI generates responses to harmfulness prompts using a helpful-only AI assistant, then critiques and revises its own responses according to a principle in the constitution, and then finetunes the original model on the revised responses.
+
+(RL Stage) AI Comparison Evaluations → Reward Model → Reinforcement Learning: The AI generates pairs of responses to harmfulness prompts using the finetuned model, then evaluates which response is better according to a principle in the constitution, and then trains a reward model from this dataset of AI preferences. The AI then trains with RL using the learned reward model.
```
thanks, fixed

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+
+.. code-block:: bash
+
+   python examples/nlp/gpt/train_reward_model.py
```
thanks, done

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+.. code-block:: bash
+
+   python examples/nlp/gpt/train_reward_model.py
+      pretrained_checkpoint.restore_from_path='/lustre/fsw/coreai_dlalgo_llm/nemo_aligner_cai/cai_tutorial/cai_mistral_7b/3_sl_cai_model/mistral_7b_sl_cai_model.nemo'
```
thanks! fixed.

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+
+.. code-block:: bash
+
+   python -u examples/nlp/gpt/train_gpt_ppo_actor.py
```
thanks, done

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+
+   python examples/nlp/cai/generate_rl_cai_dataset.py
+      --batch-size 128
+      --ngc-api-key nvapi-**_YOUR-NGC-KEY-GOES-IN-HERE_**
```
Yes. Thanks, added it.

gleibovich-nvidia commented on examples/nlp/cai/utils.py:
```diff
@@ -0,0 +1,89 @@
+from collections import defaultdict
+from multiprocessing import Pool
+
+from tqdm import tqdm
```
NeMo-Aligner is run from within the NeMo container, and 'tqdm' is present in the NeMo container's [requirements.txt](https://github.com/NVIDIA/NeMo/blob/8e65042d15062ce3fbe639f9d428c639510d894c/requirements/requirements.txt#L13)

terrykong replied:
This is more of a nit, since I highly doubt `tdqm` will ever be removed. I just wanted to point out that it's better practice to declare the packages you explicitly import and let `pip` decide if a re-install is needed. You could simply leave it unpinned in https://github.com/NVIDIA/NeMo-Aligner/blob/main/requirements.txt. What you're guarding against is if upstream repos don't use that package anymore, then your code is broken. It also makes explicit what your project expects

gleibovich-nvidia replied:
While I do agree, I would like to again stress that, as with the container, if this is something that we would like to address then this is not specific to this algorithm/tutorial, and we might want to fix it framework-wide. It is not a new requirement. Thus, I'm wondering if there was some other consideration not to add it so far. 

e.g. `tqdm` is also being used by other algorithms within NeMo-Aligner and wasn't added to the `requirements.txt`. 
https://github.com/NVIDIA/NeMo-Aligner/blob/main/examples/nlp/data/steerlm/attribute_annotate.py#L36

terrykong replied:
Since I classify this as a nit, let me just resolve this discussion as "fix later"; this shouldn't be a reason not to merge this.

gleibovich-nvidia commented on examples/nlp/cai/mistral7b_few_shot_samples:
```diff
@@ -0,0 +1,53 @@
+<s>[INST] Give me some ideas to prank my boss? [/INST]1. Change your email signature to something humorous or inappropriate.
```
We found that the few shot prompting samples significantly improve the results of the SL-CAI part of the flow (as is also mentioned in the paper).

terrykong replied:
My intention is just to reduce the risk of scrutiny by including these particular few shot examples. The options I see are:
1. just link to the few shot examples in the constitutionalAI repo
2. rewrite the prompts to be objective and without controversy
3. Just omit the few shot examples and explain to the user the format

I think (3) is the easiest, but puts the onus on the user. @gshennvm WDYT?

FWIW, (3) also has the benefit of documenting how to even create this file, which at the moment it's not clear to me how to do it given the prompt templates you're introducing and whether the special tokens need to be present in some way.

gleibovich-nvidia replied:
Oh sure, I get your point now. 

I think option (1) is less preferred as:
a. The few-shot prompts should be formatted with the matching prompt template of the LLM that we're going to align (Mistral-7B in this particular case), which the original CAI repo would be lacking.
b. It's probably better that the responses are in-distribution (on-policy), i.e. a result of generation by the LLM model to be aligned.

I'm fine with either (2) or (3), and agree regarding the benefit of (3) leading to better documentation on how to generate these.

gleibovich-nvidia replied:
I guess option (2) is a good compromise which is still run-able for the user, but I'm not entirely sure what would be the impact on the SL-CAI dataset and the trained model if the prompts and/or responses would be too neutral. We can try and rerun the whole pipeline after the change and check for the actual impact (but that might delay the release a bit). 
On the other hand, I mean, that's the whole point of this algorithm, to deal with toxic prompts/responses, so maybe it's not too bad if these examples are left as-is? 
 Anyway, I'm fine with both leaving this unchanged, or with option (2) or option (3).

gleibovich-nvidia replied:
Hi @terrykong, option (3) is implemented here: 
https://github.com/NVIDIA/NeMo-Aligner/pull/154/commits/439a389d6a6fd56c73686f5916b05b1850c3e89a

What do you think? I'm fine with it, as it allows us the flexibility we need with the few-shot samples.

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+#############################################################################
+1. Download ``Mistral-7B-Instruct`` and ``Mistral-7B`` LLM models from https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1 and https://huggingface.co/mistralai/Mistral-7B-v0.1 into the models folder.
+
+   Then convert into .nemo format:
```
Sure, I can add it. 
Although do note that other tutorials/algorithms also do not point for where to get a container. 

e.g.
https://github.com/NVIDIA/NeMo-Aligner/blob/cai/docs/user-guide/RLHF.rst?plain=1#L105
https://github.com/NVIDIA/NeMo-Aligner/blob/cai/docs/user-guide/SteerLM.rst?plain=1#L58

I'm guessing that the assumption might be that NeMo-Aligner is already run from within a NeMo container, and thus no need to point the user to where to get it from. 

@gshennvm What do you think? Is there a reason we do not specifically point the user to where to get a container from?

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+      ++model.activations_checkpoint_num_layers=1
+      ++model.optim.bucket_cap_mb=200
+      ++model.data.seq_length=4096
+      ++model.data.data_impl=json
```
thanks, fixed

gleibovich-nvidia commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,298 @@
+      ++model.optim.sched.min_lr=9e-9
+      ++model.ppo.num_rollout_samples=512
+      ++model.ppo.rollout_micro_batch_size=8
+      ++model.ppo.length_params.max_length=1024
```
thanks, fixed

gleibovich-nvidia commented on examples/nlp/cai/generate_rl_cai_dataset.py:
```diff
@@ -0,0 +1,741 @@
+\n
+"""
+
+system_prompt = f"{prefix, constitution}"
```
Thanks, renamed and fixed.

terrykong commented on examples/nlp/cai/utils.py:
```diff
@@ -0,0 +1,514 @@
+
+
+if __name__ == "__main__":
+    extra_id_prompt_template = PromptTemplate(
```
What's this main method for? It looks very nearly like what you'd write for a test. If it's just for debug, can we convert this to `utils_test.py` and use pytest to validate the templating?

gleibovich-nvidia replied:
Yes, exactly. The plan is move it to a test.

gleibovich-nvidia replied:
Fixed

terrykong commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+This will generate an SL-CAI dataset of prompts and revised responses as ``cai_revisions_aligner_chat_template.json``
+
+The few-shot samples should be provided following the template in ``few_shot_samples_example.json``, and should include a red teaming prompt, a response from the helpful model (e.g. ``Mistral-7B`` in this tutorial), critique and revision requests and responses. The Anthropic repo outlines an example `here <https://github.com/anthropics/ConstitutionalHarmlessnessPaper/blob/main/prompts/CritiqueRev ...
```
Just checking my understanding: is the user expected to only change the values corresponding to the `content`s key?

gleibovich-nvidia replied:
Yes, exactly, and potentially add several more samples if they wish (the template only include two samples, as an example). I will make this clearer.

terrykong commented:
@gleibovich-nvidia Any updates on adding some unit tests? Let me know if you have any questions

gleibovich-nvidia commented on examples/nlp/cai/generate_rl_cai_dataset.py:
```diff
@@ -0,0 +1,741 @@
+        return header
+
+    @staticmethod
+    def _apply_role_template(role: str, prompt: Optional[str] = None):
```
moved to `utils/cai_utils` and added pytest tests for all of our use-cases.

gleibovich-nvidia commented:
> @gleibovich-nvidia Any updates on adding some unit tests? Let me know if you have any questions

Sure, just added a module `tests/test_cai_utils.py` to test for extra utils functionality that this PR has added.

gleibovich-nvidia commented:
> Hi @gleibovich-nvidia ! Thanks so much for this contribution, and expanding our supported algorithms.
> 
> Apologies in advance since my review is as large due to the large PR
> 
> The documentation is great.
> 
> I do think it'd be helpful to add tests for the scripts you added, in particular:
> 
> * `generate_rl_cai_dataset.py` -> `generate_rl_cai_dataset_test.py`
> * `generate_sl_cai_dataset.py` -> `generate_sl_cai_dataset_test.py`
> * `process_anthropic_hh_using_chat_prompt.py` -> `process_anthropic_hh_using_chat_prompt_test.py`
>   where the test adds a pytest that shows a valid input and expected output. My understanding is that these scripts also rely on a remote inference call, so in the test, it's best to just mock that out.
> 
> I also saw a few hard coded `extra_id_*` and I think those should be cleaned up so that it is configurable and plumbed from the user; especially since other scripts allow those to be configured to some degree.

Hi @terrykong, thanks so much for the review and excellent feedback! 
I think it helped increasing the PR quality quite significantly. 

At this point, I think that other than 3 additional comments requiring @gshennvm attention, we have addressed and fixed all the open issues. 

As for e2e tests for `generate_rl_cai_dataset.py`, `generate_sl_cai_dataset.py` and `process_anthropic_hh_using_chat_prompt.py`, let's discuss what exactly we want to get done here. 

As CAI is not a new distinct algorithm within NeMo-Aligner, and is more of a tutorial, I think (and also following a short chat with @gshennvm on this matter), we should focus on testing code functionality which is dependent on NeMo-Aligner's APIs. This is in order to make sure not to break in case one of those APIs introduce breaking changes. This is consistent with how other tutorials are implemented in the framework (e.g. SteerLM and DRaFT+), which do not implement e2e tests on their own. 

Reviewing those specific scripts, the only dependency is on `cai_utils` for which we have introduced `pytest` tests as part of this PR. There are no NeMo-Aligner APIs dependencies, as running SFT, RM and PPO is executed by the user (cmdline runs), following and in-between SL-CAI and RL-CAI datasets generation. 

 I would be happy to get your feedback and thoughts on how to proceed, and what we want to get done to finalize the last steps required to get this PR merged. 

Thanks! 
Gal

terrykong commented on nemo_aligner/utils/cai_utils.py:
I feel that this should remain in examples/nlp/cai since the pattern thus far has been to leave utils similar to these in their respective examples folder:
* https://github.com/NVIDIA/NeMo-Aligner/blob/5b9c1f9aaeb08894be1c05bb5c9f673e8e1b4bf7/examples/nlp/data/steerlm/preprocess_helpsteer_data.py#L23
* https://github.com/NVIDIA/NeMo-Aligner/blob/5b9c1f9aaeb08894be1c05bb5c9f673e8e1b4bf7/examples/nlp/data/steerlm/preprocess_openassistant_data.py#L63

We can consider promoting it into the main package once another project requires these exact utils.

gleibovich-nvidia replied:
Agree. The reason I moved it to `utils` is to allow for "easy" importing from `tests/test_cai_utils.py`. It is not importable otherwise, as `examples` is not part of the package. We also can't have `test_cai_utils.py` under `examples/nlp/cai` as pytest would ignore it.

There are thus three ways to resolve:
1. Have `cai_utils` under `utils` (as is it done now).
2. Add `__init__.py` to `examples` and `examples/nlp`. 
3. Add the path to dynamically to `sys.path`. 

I didn't like options (2) and (3) so much as option (2) enforces `examples` to be part of the package, which I'm not sure we want. Option (3) is also not ideal since messing around with `sys.path` might lead to unexpected hard to debug issues on the user side. 

Since, as I understand, option (1) is also less preferred, we can go with option (3). I have pushed this change now, PTAL and let me know what you think. If you have other ideas in mind, please let me know and I would be happy to follow up and implement.

terrykong commented:
@gleibovich-nvidia @snisimov Thanks for addressing and iterating on the PR. I have also asked our writer to review the docs as these will surface in the Nemo FW Docs

CC @jgerh

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+`Constitutional AI (CAI)`_ is an approach by Anthropic to train AI systems that are helpful, honest, and harmless, without relying on human feedback labels for harmlessness.
+CAI uses a set of natural language principles to guide AI behavior and self-improvement.
+The method consists of two stages: a supervised learning stage, where the AI critiques and revises its own responses according to the principles, and a reinforcement learning stage, where the AI learns from its own feedback based on the principles.
+CAI allows training a harmless but non-evasive AI assistant that engages with harmful queries by explaining its objections to them.
```
harmless, but

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+CAI
+###############
+The basic steps of CAI, as illustrated in Figure 1 of the paper (below):
```
The basic steps of CAI are described in this section and illustrated in the figure below.

gleibovich-nvidia replied:
Thanks, revising to: 
"The basic steps of CAI are described in this section and illustrated in the figure below (Figure 1)."

jgerh replied:
There are two figures in this file. If you want to refer to a figure by a number, you need to edit the figure title and add Figure 1. "<Title>" and Figure 2. "<Title>". I suggest  not adding figure numbers in this file as there are only two figures in this file and it is not necessary. 

Instead, retain my original edit: The basic steps of CAI are described in this section and illustrated in the figure below.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+###############
+The basic steps of CAI, as illustrated in Figure 1 of the paper (below):
+
+(Supervised Stage) Critique → Revision → Supervised Learning: The AI generates responses to harmfulness prompts using a helpful-only AI assistant, then critiques and revises its own responses according to a principle in the constitution, and then finetunes the original model on the revised responses.
```
fine-tunes

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+.. image:: ../assets/cai_diagram.png
+   :alt: basic steps of the CAI process
+
+Critiques, revisions and AI harmlessness feedback are steered by a small set of principles drawn from a ‘constitution’. The supervised stage significantly improves the initial model, and gives some control over the initial behavior at the start of the RL phase, addressing potential exploration problems. The RL stage significantly improves performance and reliability.
```
'constitution.'

gleibovich-nvidia replied:
?
Why move the period inside the quotes?

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+.. image:: ../assets/cai_diagram.png
+   :alt: basic steps of the CAI process
+
+Critiques, revisions and AI harmlessness feedback are steered by a small set of principles drawn from a ‘constitution’. The supervised stage significantly improves the initial model, and gives some control over the initial behavior at the start of the RL phase, addressing potential exploration problems. The RL stage significantly improves performance and reliability.
```
The supervised stage significantly improves the initial model.  It gives some control over the initial behavior at the start of the RL phase, while addressing potential exploration problems.

jgerh replied:
revisions, and

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+Motivation
+###############
+
```
Constitutional AI motivation refers to designing AI systems in such a way that their objectives and behaviors are guided by a set of predefined rules or principles. It includes the following:

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+2. Generate responses to harmfulness prompts using a helpful-only AI assistant. Ask the model to critique its response according to a principle in the constitution, and then revise the original response in light of the critique.
+
+3. Finetune ``Mistral-7B`` with SFT on the revised responses to create a ``Mistral-7B-SL-CAI`` model.
```
Fine-tune

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+3. Finetune ``Mistral-7B`` with SFT on the revised responses to create a ``Mistral-7B-SL-CAI`` model.
+
+4. 
```
Delete 4.

gleibovich-nvidia replied:
Steps 3 and 4 are two distinct stages. I have added the missing headline for steps a. b. and c. to step 4.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+3. Finetune ``Mistral-7B`` with SFT on the revised responses to create a ``Mistral-7B-SL-CAI`` model.
+
+4. 
+   a. Use ``Mistral-7B-SL-CAI`` to generate a set of candidate responses to each prompt in a dataset of toxic prompts.
```
replace a. with a bullet character and indent text under step 3.

jgerh replied:
Lines 54 to 57. Changes are good.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+4. 
+   a. Use ``Mistral-7B-SL-CAI`` to generate a set of candidate responses to each prompt in a dataset of toxic prompts.
+   b. Formulate each prompt and pair into a multiple choice question, where we ask ``Mixtral-8x7B`` which response is best according to the constitution.
```
replace b. with a bullet character and indent text under step 3.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+4. 
+   a. Use ``Mistral-7B-SL-CAI`` to generate a set of candidate responses to each prompt in a dataset of toxic prompts.
+   b. Formulate each prompt and pair into a multiple choice question, where we ask ``Mixtral-8x7B`` which response is best according to the constitution.
+   c. Blend the AI feedback preference dataset (prompts and pairs) with human feedback helpfulness dataset.
```
replace c. with a bullet character and indent text under step 3.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+   b. Formulate each prompt and pair into a multiple choice question, where we ask ``Mixtral-8x7B`` which response is best according to the constitution.
+   c. Blend the AI feedback preference dataset (prompts and pairs) with human feedback helpfulness dataset.
+
+5. Train a RM.
```
Replace RM with Reward Model (RM).

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+5. Train a RM.
+
+6. Finetune the ``Mistral-7B-SL-CAI`` with PPO and the RM to train a ``Mistral-7B-RL-CAI`` model.
```
Fine-tune
replace PPO with Proximal Policy Optimization (PPO)

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+6. Finetune the ``Mistral-7B-SL-CAI`` with PPO and the RM to train a ``Mistral-7B-RL-CAI`` model.
+
+7. Inference.
```
7. Run inference.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+(Supervised Stage) Critique → Revision → Supervised Learning: The AI generates responses to harmfulness prompts using a helpful-only AI assistant, then critiques and revises its own responses according to a principle in the constitution, and then finetunes the original model on the revised responses.
+
+(RL Stage) AI Comparison Evaluations → Reward Model → Reinforcement Learning: The AI generates pairs of responses to harmfulness prompts using the finetuned model, then evaluates which response is better according to a principle in the constitution, and then trains a reward model based on this dataset of AI preferences and a human helpfulness preferences. The AI then trains with RL using the lear ...
+
```
Add a figure title: Constitutional AI Basic Steps

gleibovich-nvidia replied:
rst doesn't support figure titles. I have tried several workaround but failed getting it to render properly. I think that the first sentence in this paragraph is sufficient. thanks.

jgerh replied:
OK, I understand. If we've run into the limitations of rst, then I would suggest removing the figure number altogether and just state "the figure below" or "the following figure." The same suggestion for the other figure in this file. That should work.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+#############################################################################
+1. Download ``Mistral-7B-Instruct`` and ``Mistral-7B`` LLM models from https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1 and https://huggingface.co/mistralai/Mistral-7B-v0.1 into the models folder.
+
+   Then convert into .nemo format:
```
Then,

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+      python3 -c "from datasets import load_dataset; load_dataset('Anthropic/hh-rlhf', data_dir='red-team-attempts', split='train').to_json('/path/to/anthropic_red_team_attempts_train.json')"
+   
+   This will download the dataset to ``/path/to/anthropic_red_team_attempts_train.json``
```
This command will download the dataset to ``/path/to/anthropic_red_team_attempts_train.json``.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+   
+      python3 -c "from datasets import load_dataset; load_dataset('nvidia/sft_datablend_v1', split='train').to_json('/path/to/nvidia_sft_datablend_v1_train.json')"
+
+   This will download the dataset to ``/path/to/nvidia_sft_datablend_v1_train.json``
```
This command will download the dataset to ``/path/to/nvidia_sft_datablend_v1_train.json``.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+The few-shot samples should be provided following the template in ``few_shot_samples_example.json`` (filling in the `content` tags, and choosing how many samples to use), and should include a red teaming prompt, a response from the helpful model (e.g. ``Mistral-7B`` in this tutorial), critique and revision requests and responses. The Anthropic repo outlines an example `here <https://github.com/an ...
+
+*NOTE: The tokenizer file can be found by extracting the .nemo checkpoint using `tar -xf /models/mistral/mistral-7b-Instruct.nemo`.
+There are 2 tokenizer files that end with `*.model` in the model checkpoint and they are the same, so you can use either one for data processing.*
```
Delete * after data processing.* It shows up in the rendered text. I don't think it is needed, but pls check.

gleibovich-nvidia replied:
Thanks, fixed.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+This will generate an SL-CAI dataset of prompts and revised responses as ``cai_revisions_aligner_chat_template.json``
+
+The few-shot samples should be provided following the template in ``few_shot_samples_example.json`` (filling in the `content` tags, and choosing how many samples to use), and should include a red teaming prompt, a response from the helpful model (e.g. ``Mistral-7B`` in this tutorial), critique and revision requests and responses. The Anthropic repo outlines an example `here <https://github.com/an ...
```
Change "The Anthropic repo outlines an example [here](https://github.com/anthropics/ConstitutionalHarmlessnessPaper/blob/main/prompts/CritiqueRevisionFewShotPrompts.json)" to "An example is shown in the Anthropic repo here:"

gleibovich-nvidia replied:
The word 'here' is already a hyperlink.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+Step 3: Finetune Mistral-7B on the revised responses to create a Mistral-7B-SL-CAI model
+######################################################################################################
+
+Note that you would need to set up multi-node training run in your cluster env, depending on the type of cluster you use. For details, please refer to https://lightning.ai/docs/pytorch/stable/clouds/cluster.html
```
Add a period at the end of the sentence: For details, please refer to https://lightning.ai/docs/pytorch/stable/clouds/cluster.html.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+##############################################################################################################
+
+The following section runs an inference server with the SL-CAI model that we've previously trained, and queries it with red teaming prompts asking for several responses per prompt.
+These will then be ranked by a judge LLM being run from NVIDIA's NGC. An NGC API key can be acquired `here`_.
```
The prompts will then be ranked by a judge LLM being run from NVIDIA's NGC.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+The following section runs an inference server with the SL-CAI model that we've previously trained, and queries it with red teaming prompts asking for several responses per prompt.
+These will then be ranked by a judge LLM being run from NVIDIA's NGC. An NGC API key can be acquired `here`_.
+
+The following will run the inference server:
```
The following commands will run the inference server:

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+      --blend-with "{'name': 'anthropic_helpful_only', 'train': {'prompts': ['/path/to/anthropic_helpful_only/anthropic_helpful_only_train_prompts_with_chat_prompt.jsonl'], 'comparisons': ['/path/to/anthropic_helpful_only/anthropic_helpful_only_train_comparisons_with_chat_prompt.jsonl']}, 'test': {'prompts': ['/path/to/anthropic_helpful_only/anthropic_helpful_only_test_prompts_with_chat_prompt.js ...
+      --port 5999
+
+This will create the ``rl-cai`` dataset files in the defined output folder with the given output filename prefix.
```
This command will create the  . . .

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+Step 5: Train the RM
+#####################
+
```
The following command will train the RM:

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+
+The trained RM checkpoint will be saved to output dir given by ``exp_manager.explicit_log_dir``.
+
+Step 6: Finetune Mistral-7B-SL-CAI with PPO and the RM to train a Mistral-7B-RL-CAI model
```
Fine-tune

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+      model.seed=1234 \
+      exp_manager.explicit_log_dir=<path to critic output dir>
+
+Run the following command to launch actor training and reference policy server:
```
and a reference policy server:

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+      remote_critic_rm.critic.ip=<ip to critic service> \
+      remote_critic_rm.critic.port=5567
+
+The trained policy (LLM) checkpoint will be saved to output dir given by ``exp_manager.explicit_log_dir``.
```
The trained LLM policy checkpoint will be saved to the output dir given by . . .

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,308 @@
+           web_server=False \
+           port=1427
+
+Please wait for the server to be ready before proceeding, and follow the on-screen directive.
```
Please wait for the server to be ready before proceeding. Then, follow the on-screen directive.

jgerh commented on docs/user-guide/index.rst:
```diff
@@ -20,6 +20,9 @@
    DRaFT+ is an algorithm for fine-tuning text-to-image generative diffusion models by directly backpropagating through a reward model which alleviates the mode collapse issues from DRaFT algorithm and improves diversity through regularization. 
 
+:ref:`Constitutional AI: Harmlessness from AI Feedback <model-aligner-cai>`
+   CAI is an alignment method by Anthropic allowing to incorporate AI feedback for LLM model alignment. The feedback is based on a small set of principles (Constitution) guiding the model towards the desired behavior - being helpful, honest and harmless.
```
CAI is an alignment method by Anthropic allowing it to incorporate AI feedback for LLM model alignment.

gleibovich-nvidia replied:
I'm not sure what does "it" is referring to here, but maybe I'm misunderstanding. 

CAI is a method which allows to incorporate AI feedback into model alignment using NeMo-Aligner. The method was invented by Anthropic. How about the following?

"CAI is an alignment method by Anthropic allowing incorporation of AI feedback for LLM model alignment

jgerh replied:
Yes: "CAI is an alignment method by Anthropic allowing incorporation of AI feedback for LLM model alignment."

jgerh commented on examples/nlp/cai/sys_prompt_constitution:
```diff
@@ -0,0 +1,54 @@
+
+You will be given a prompt and two or more responses to this prompt.
+Your task is to select exactly 2 responses:
+1. select the response which is most aligned with the principles in the CONSTITUTION. label this response as "chosen".
```
1. Select the response which is most aligned with the principles in the CONSTITUTION. Label this response as "chosen."

jgerh commented on examples/nlp/cai/sys_prompt_constitution:
```diff
@@ -0,0 +1,54 @@
+You will be given a prompt and two or more responses to this prompt.
+Your task is to select exactly 2 responses:
+1. select the response which is most aligned with the principles in the CONSTITUTION. label this response as "chosen".
+2. select the response which is the least aligned with the principles in the CONSTITUTION. label this response as "rejected"
```
2. Select the response which is the least aligned with the principles in the CONSTITUTION. Label this response as "rejected."

jgerh reviewed (commented):
I completed the copyedits to the rst files identified in [Constitutional AI: Harmlessness from AI Feedback](https://github.com/NVIDIA/NeMo-Aligner/pull/154/files#top)
#154 and submitted them for your review. Please let me know if you have any questions.

terrykong commented:
@gleibovich-nvidia Please review @jgerh 's copyedits

gleibovich-nvidia commented:
> I completed the copyedits to the rst files identified in [Constitutional AI: Harmlessness from AI Feedback](https://github.com/NVIDIA/NeMo-Aligner/pull/154/files#top) #154 and submitted them for your review. Please let me know if you have any questions.

Thanks a lot! 
I have accepted and incorporated most of the suggestions. There are couple of comments, which I'm not sure about, to which I've replied directly.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,313 @@
+
+CAI
+###############
+The basic steps of CAI are described in this section and illustrated in the figure below (`Figure 1 <https://arxiv.org/abs/2212.08073>`_).
```
The basic steps of CAI are described in this section and illustrated in `Figure 1 <https://arxiv.org/abs/2212.08073>`_).

gleibovich-nvidia replied:
I think I didn't explain myself well enough in the previous discussion, sorry about that. 

By referring to Figure 1, I'm not referring to Figure 1 in this specific document, but am referring to Figure 1 from CAI paper by anthropic. I.e. this is more of a citation, trying to make sure that the user understands that this is not a figure that we've made, but is taken from the paper.

See Figure 1 here. - https://arxiv.org/pdf/2212.08073.
<img width="640" alt="image" src="https://github.com/NVIDIA/NeMo-Aligner/assets/150582838/d07b3d8e-91bc-44d4-8c9e-8a3bd904bed6">

How about the following to make this clearer:
The basic steps of CAI are described in this section and illustrated in the figure below (`Figure 1 <https://arxiv.org/abs/2212.08073>`_, Bai et. al).

jgerh replied:
OK, thank you for the clarification. The text as written is correct then.

jgerh commented on docs/user-guide/CAI.rst:
```diff
@@ -0,0 +1,313 @@
+.. image:: ../assets/cai_diagram.png
+   :alt: basic steps of the CAI process
+
+   Constitutional AI Steps. `Figure 1 <https://arxiv.org/abs/2212.08073>`_.
```
`Figure 1. Constitutional AI Steps <https://arxiv.org/abs/2212.08073>`_.

gleibovich-nvidia commented:
@terrykong FYI, we found an issue with the converted tokenizer being used within `megatron_gpt_eval.py` causing the BOS token of Mistral's prompt template not be handled correctly, when used in a conversation, rather than in a single prompt/reply. This is especially important when using few shot prompting, as the model observes an already in progress conversation (which should follow the template correctly). It seems that `megatron_gpt_eval.py` is not a good fit here as an API inference service for a conversation (chat) with several rounds of human/assistant interaction, as it does not support using HF's tokenizer natively. 

To resolve this issue, commit https://github.com/NVIDIA/NeMo-Aligner/commit/65a9c3b3531b43cee04c070185268465bcae123f is adding a lightweight inference service overriding the tokenizer to use the HF tokenizer instead, which tokenizes a conversation correctly. 
NeMo's documentation is also mentioning using the HF tokenizer [here](https://docs.nvidia.com/nemo-framework/user-guide/latest/llms/mistral/trainingpredefined.html).

terrykong commented:
> we found an issue with the converted tokenizer being used within megatron_gpt_eval.py causing the BOS token of Mistral's prompt template not be handled correctly

Is my understanding correct that the tokenizer within the megatron checkpoint does not tokenize the BOS as you expect?

Why not change the prompt template to use the BOS token that the tokenizer expects? I feel like maintaining a light fork of the megatron_gpt_eval.py should be a last resort.

jgerh commented:
@gleibovich-nvidia Gal, I reviewed your comments, responded to your questions, and completed my review.

gleibovich-nvidia commented:
> > we found an issue with the converted tokenizer being used within megatron_gpt_eval.py causing the BOS token of Mistral's prompt template not be handled correctly
> 
> Is my understanding correct that the tokenizer within the megatron checkpoint does not tokenize the BOS as you expect?
> 
> Why not change the prompt template to use the BOS token that the tokenizer expects? I feel like maintaining a light fork of the megatron_gpt_eval.py should be a last resort.

Yes, the BOS for Mistral models is `<s>` and it is then tokenized to three different tokens `<`, `s`, `>` with the tokenizer in the megatron checkpoint, instead of tokenizing it to just one single token `<s>`. 

We prompt a **trained instruct** model (`Mistral-7B-v0.1`) asking for critiques and revisions. The model then expects this specific prompt template, considering the way it was trained and the data it has observed during SFT.

While I do agree it is not ideal to maintain code adapted from NeMo, I would like to stress that:
1.  `megatron_gpt_eval.py` is merely an example of how to run inference with `NeMo`. AFAIU, it wasn't developed to be a comprehensive solution for LLM inference, but just an example of how it can be done for specific use-cases. It doesn't implement any combination of use-cases required by potential users. This PR, on the other hand, is requiring functionality which is missing with `megatron_gpt_eval.py` to get a prompt tokenized correctly to match Mistral's expected prompt template. 
2. There are other examples within `NeMo-Aligner` where missing upstream functionality required some code adaptation and duplication. e.g. https://github.com/NVIDIA/NeMo-Aligner/blob/main/nemo_aligner/utils/train_utils.py#L16 and https://github.com/NVIDIA/NeMo-Aligner/blob/main/nemo_aligner/data/nlp/samplers.py#L28. 

There might be another way to workaround this issue that I can try on Sunday morning. Let's see how that goes. 
If all else fails, we can stick with this proposed workaround for now, and I will make sure to add a note highlighting that this code is adapted from NeMo, and we can remove this extra added functionality once this issue gets resolved in NeMo.

gleibovich-nvidia commented:
> @gleibovich-nvidia Gal, I reviewed your comments, responded to your questions, and completed my review.

Thanks a lot @jgerh!

jgerh reviewed (dismissed):
Completed reviews and updates from Gal Leibovich.

terrykong commented:
> Yes, the BOS for Mistral models is \<s\> and it is then tokenized to three different tokens \<, s, \> with the tokenizer in the megatron checkpoint, instead of tokenizing it to just one single token \<s\>.
>
> We prompt a trained instruct model (Mistral-7B-v0.1) asking for critiques and revisions. The model then expects this specific prompt template, considering the way it was trained and the data it has observed during SFT.

I feel like I'm missing something b/c shouldn't [convert_mistral_7b_hf_to_nemo.py](https://github.com/NVIDIA/NeMo/blob/main/scripts/checkpoint_converters/convert_mistral_7b_hf_to_nemo.py) add the tokenizer that the huggingface checkpoint was trained with? If for some reason you need a different one, I feel like the functionality should just be added to `convert_mistral_7b_hf_to_nemo.py`

> megatron_gpt_eval.py is merely an example of how to run inference with NeMo. AFAIU, it wasn't developed to be a comprehensive solution for LLM inference, but just an example of how it can be done for specific use-cases. It doesn't implement any combination of use-cases required by potential users. This PR, on the other hand, is requiring functionality which is missing with megatron_gpt_eval.py to get a prompt tokenized correctly to match Mistral's expected prompt template.

That's true that one script may not be comprehensive, but this example seems to just be about a tokenizer being different, so I feel like that's not a big enough reason to fork. Tokenizers are usually part of the package deal for models since swapping tokenizers can usually result in garbage when the token-ids get remapped, so I prefer to address this in the model packaging/conversion rather at inference time when swapping the tokenizer is not so much a feature, but an avenue for user error.

> There are other examples within NeMo-Aligner where missing upstream functionality required some code adaptation and duplication. e.g. https://github.com/NVIDIA/NeMo-Aligner/blob/main/nemo_aligner/utils/train_utils.py#L16 and https://github.com/NVIDIA/NeMo-Aligner/blob/main/nemo_aligner/data/nlp/samplers.py#L28.

Sure, but these are not examples to follow, but exceptions. Also, the second one even says it should be deleted.


In general, forking puts the burden on us to maintain the functionality and correctness. I would rather us focus on the algorithms here and push for Nemo/Megatron to incorporate our features when possible. If we ask and they say "no" or "not anytime soon", then I see that as a good reason for when we should fork.

terrykong commented on docs/user-guide/CAI.rst:
Can you make this lower case? There is a new requirement that restructured text docs have to have lower case filenames

gleibovich-nvidia replied:
Done

jgerh replied:
IPlease check the index.rst file here https://github.com/NVIDIA/NeMo-Aligner/blob/f60ea0a6b05892da23f852c4f1dd6864cd00eeb5/docs/user-guide/index.rst and change CAI.rst to cai.rst.
![cai](https://github.com/NVIDIA/NeMo-Aligner/assets/163925524/97210dae-7ae8-40ac-9096-f0c74418e7d0)

gleibovich-nvidia replied:
Thanks! I've missed this one. Fixed.

jgerh commented on docs/user-guide/index.rst:
```diff
@@ -13,6 +13,7 @@
+   cai.rst
 
 :ref:`Prerequisite Obtaining a Pre-Trained Model <prerequisite>`
    This section provides instructions on how to download pre-trained LLMs in .nemo format. The following section will use These base LLMs for further fine-tuning and alignment. 
```
This section provides instructions on how to download pre-trained LLMs in .nemo format. The following section will use these base LLMs for further fine-tuning and alignment.

snisimov replied:
These are parts of the document not related to CAI, i think it would be better to open a separate issue.

terrykong replied:
@jgerh Can you open a separate PR to address these documentation changes?

jgerh commented on docs/user-guide/index.rst:
```diff
@@ -34,3 +35,5 @@
 :ref:`Fine-tuning Stable Diffusion with DRaFT+ <model-aligner-draftp>`
    DRaFT+ is an algorithm for fine-tuning text-to-image generative diffusion models by directly backpropagating through a reward model which alleviates the mode collapse issues from DRaFT algorithm and improves diversity through regularization. 
+:ref:`Constitutional AI: Harmlessness from AI Feedback <model-aligner-cai>`
+   CAI is an alignment method by Anthropic allowing incorporation of AI feedback for LLM model alignment. The feedback is based on a small set of principles (Constitution) guiding the model towards the desired behavior - being helpful, honest and harmless.
```
Suggested revision:

CAI, an alignment method developed by Anthropic, enables the incorporation of AI feedback for aligning LLMs. This feedback is grounded in a small set of principles (referred to as the ‘Constitution’) that guide the model toward desired behaviors, emphasizing helpfulness, honesty, and harmlessness.

snisimov replied:
thank you we will change.

terrykong commented:
Needs https://github.com/NVIDIA/NeMo-Aligner/pull/229

gleibovich-nvidia commented:
> Needs #229

Merged with `main`, and tested to work well. I think we're ready to merge.

gleibovich-nvidia commented on examples/nlp/cai/generate_rl_cai_dataset.py:
```diff
@@ -0,0 +1,741 @@
+    @staticmethod
+    def extract_response(prompt: str):
+        response = prompt.rsplit(ChatPromptTemplate.assistant_message_header, 1)[-1]
+        response = response.strip().removesuffix(ChatPromptTemplate.turn_token).strip()
```
nemo requires python >= 3.10

terrykong reviewed (approved):
Thanks @gleibovich-nvidia and @snisimov for working thru the review and adding CAI! LGTM!
